FHIR/fhir-package-loader
# FHIR Package Loader Rewrite

**Description:** _And now for something completely different..._

This PR represents a total rewrite of the FHIR Package Loader -- with completely new interfaces, implementations, and APIs. This rewrite was necessary to address limitations in the package loader as well as support future use cases (such as transitive dependencies, multiple versions of the the same package, etc).

The new FHIR Package Loader now utilizes an in-memory database (currently [sql.js](https://www.npmjs.com/package/sql.js)) to store and query metadata about packages and their resources. This reduces memory usage per-resource since now only the metadata is kept in memory (vs the entire JSON object). With more efficient memory usage, we can now load all resources from the package (instead of just conformance resources) and we will also be able to load transitive packages in the future.

Separate concerns of the package loader have been separated out into their own interfaces and implementations, including:
* `PackageDB` interface w/ `SQLJSPackageDB` implementation. Note that sql.js was chosen due to its support for synchronous queries -- which causes the least disruption to SUSHI (which has many synchronous functions).
* `RegistryClient` interface w/ `FHIRRegistryClient`, `NPMRegistryClient`, `RedundantRegistryClient`, and `DefaultRegistryClient` implementations
* `CurrentBuildClient` interface w/ `BuildDotFhirDotOrgClient` implementation
* `PackageCache` interface w/ `DiskBasedPackageCache` implementation
* `VirtualPackage` interface w/ `DiskBasedVirtualPackage` and `InMemoryVirtualPackage` implementations
* `PackageLoader` interface w/ `BasePackageLoader` implementation that accepts implementations of the previous interfaces and ties them together for a complete package loading solution.

This approach allows for additional capabilities in the future, such as implementations that work well in web browsers.

**Testing Instructions:**
* Ensure that all unit tests pass
* Try the command-line client as described in the README
* Try [SUSHI's new-fpl branch](https://github.com/FHIR/sushi/tree/new-fpl) that integrates this PR's package loader

_NOTE: Much of this PR contains code that was previously reviewed before merging into the `new-fpl` feature branch. Those reviews, however, typically only involved one reviewer -- and some of the initial code never went through a PR. So... review and test accordingly._

**Related Issue:** Implements #30.
